README.md
# EFEMem DB

**Easy, Fast and Efficient MEMory NoSQL DataBase**

**version 1.0.8**

<img src="EFEMemDB-icon.png" alt="EFEMem NoSQL Database" style="zoom:50%;" />

*Created by Rafael Hernamperez and released under the terms of the ISC License:*

https://opensource.org/licenses/ISC



**ISC License (ISC)**
**Copyright 2020 Rafael Hernamperez**

Permission to use, copy, modify, and/or distribute this software for any purpose with or without fee is hereby granted, provided that the above copyright notice and this permission notice appear in all copies.

THE SOFTWARE IS PROVIDED "AS IS" AND THE AUTHOR DISCLAIMS ALL WARRANTIES WITH REGARD TO THIS SOFTWARE INCLUDING ALL IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS. IN NO EVENT SHALL THE AUTHOR BE LIABLE FOR ANY SPECIAL, DIRECT, INDIRECT, OR CONSEQUENTIAL DAMAGES OR ANY DAMAGES WHATSOEVER RESULTING FROM LOSS OF USE, DATA OR PROFITS, WHETHER IN AN ACTION OF CONTRACT, NEGLIGENCE OR OTHER TORTIOUS ACTION, ARISING OUT OF OR IN CONNECTION WITH THE USE OR PERFORMANCE OF THIS SOFTWARE.



## Description

**EFEMem DB** is an ultra-light, fast, easy and efficient NoSQL database, based on key-value pairs, running entirely on RAM memory. 

**EFEMem DB** is written in **JavaScript**, and you can run it directly on the front size, such us web browsers, through web applications (**web pages**. **PWA** (Progressive Web Applications), hybrid mobile applications (made with **Ionic**, **React Native**, **NativeScript**, **Xamarin**, etc.). You can control ad-hoc data without APIs, speeding up your applications. **EFEMem DB** is an ideal solution for your data when the connection is off, providing you a high availability for your users.

You can also run **EFEMem DB** on server side applications, using technologies based on JavaScript, such us **NodeJS**. You can achieve more performance managing master data, cache data, log or monitoring data, highly queried data... Avoid heavy, slow queries and repeated queries to remote databases, optimizing the time and efficiency of your most used data or managing data in real-time.

With a few and very easy commands, you can take the control of your data.

```html
// Web application example
<script src="uri/efememdb.js"></script>

<script>
   let result = efemem.set("hello", "world");
   result = efemem.get("hello");
</script>
```



```javascript
// NodeJS example
const { efemem } = require('efememdb');

let result = efemem.set("hello", "world");
result = efemem.get("hello");
```





## Features

These are the main features of **EFEMem DB**:

- **Easy to install.** 
- **No third-party dependencies.** All the code is own.
- **Ultra-fast read and write access** (few nanoseconds).
- **Direct usage on the scope of your web application**, without calls to APIs.
- **Performance improvement on your server applications,** accessing fastly to your most frequently data or managing data on real time.
- **Hot configuration.** You can change configuration while **EFEMem DB** is running.
- **Recycling mode data.** You can use this mode in order to contain a maximum number of keys. For example, you can store the 1000 last logs or requests or any other data. When the space is full, the next key will be added to the bottom after the first key (the oldest) is deleted.
- **Life time key.** You can define the life time of a given key. For example, you can create a key that expires in 30 seconds.
- **Powerful commands.** You can control the data with a few and easy commands.
- **Persistence.** The data can be permantely saved (into **files** disk on server side, or into **localStorage** browser), and restore it when you want.
- **NoSQL.** You don't need complex relationships between tables, or strict schemas. You use key-value pairs.
- **Space names.** You can organize your keys using space names. You can use keys with the same name, but unique in different spaces.
- **Pattern names.** You can access efficiently to many keys and spaces using pattern names, instead of accessing one by one.
- **Rich data.** You can use different types of data:
  - integer numbers
  - float numbers
  - string text
  - boolean
  - arrays
  - JSON objects



### Best usage

#### Server side

You can use this amazing data base in your **NodeJS**, as a complementary database for highly and immediate availability of data, such as real-time monitoring, recycled logs, cache, configuration, master data, etc. It's ideally for the following cases:

- Master data. Avoid constant queries to master data
- Data in real-time: Monitoring.
- Cycling logs.
- Frequently data accessed. Caches.
- IoT projects
- Temporary data



#### Front side

You can use **EFEMem DB** in your web applications, without need to call remote, heavy and slow APIs, storing and retrieving data for local web purposes. It's ideally for the following cases:

- Data management when the connection is off, assuring the high availability of your app until the connection is recovered. This feature is ideal for mobile applications, PWAs and web applications in general.
- Local configuration storage that can be changed.
- Sessions management.
- Local data managed exclusively by your web app.
- HTML games
- Temporary data





## Installation

### NodeJS

The recommended installation of **EFEMem DB** is via `npm` installation package:

```javascript
npm install efememdb -save
```



### Web applications

Copy the file `efememdb.js` on your project. This file contains itself all the functionalities that are required by **EFEMem DB**.





## First steps

Once you installed **EFEMem DB** in your project, the first step is import the `efemem` object.

If you are using **EFEMem DB** on NodeJS:

```javascript
const { efemem } = require('efememdb');
```



If you are using **EFEMem DB** on a web project:

<script src="path/efememdb.js"></script>



The `efemem` object allows you to execute directly the **EFEMem DB** commands, through which you can interact with the data.

```javascript
let result = efemem.command(<parameters>)
```



Each command returns a result, commonly with this format:

```json
{
  "ok": true|false,
  "cmd": "command",
  "data": data,
  "msg": "result_message",
  "affected": num_affected,
  "time": "execution_time"
}
```



Where:

- **`ok`**: Returns true if the command has been executed successfully. Returns false if any issue of error has been detected in the execution.
- **`cmd`**: Name of the executed command
- **`data`**: Data used or result data (depending on the command)
- **`msg`**: Message or description about the result.
- **`affected`**: Number of items affected by the command (0 if error or incident)
- **`time`**: Time used by the command in its execution



**Note:** *Depending of the command, this format can vary.* 





## Concepts

### Spaces

The spaces organize or group the keys under a single and common name, in order to facilitate the search and storage ot the values.



### Keys

One key identify in explicity way a given value.

The name of a key is unique in the space name in which is stored. You can define keys with the same name, but unique on its space name.

You will use the key in order to access to the values. For this reason, the name of the key must be powerfully significant and relevant. 

You can use simple names, or, depending on the number and complexity of data structure, you can use composed names, using special characters (like colon, pipe or number), in order to organize the parts of the key name.

For example, this key:

`student:001|city:002|course:012`

contains relevant informartion in order to associate and find the value:

- Student ID = 001
- City ID = 002
- Course ID = 012



### Values

Each value is associated to an unique key. You can use different types of values:

- numbers
- text
- boolean
- array
- JSON objects



### Pattern names

Pattern name is a powerful feature of **EFEMem DB** than allows you to spread the scope of results of some commands. When you want to refer to a key or a space name, you can specify a part of the name, instead of the complete name. This part can be located at the begining, in the middle or at the end of the name (in any part of the name). If you don't specify any pattern (null or empty value), **EFEMem DB** will assume any name (all names).

For example, you can save the info of the courses on spaces with their corresponding year: `course2016`, `course2017`, `course2018`, `course2019` and `course2020`. You can execute a command than affect to all the courses of the 10's decade, using as pattern name `'course201'` (this excludes `course2020`).



### Define keys and values

You can define a key using this command:

```javascript
efemem.set("language:en", "English");
efemem.set("language:es", "Spanish");
```



By default, **EFEMem DB** uses the space name `'public'` in order to store the keys.

A good practice is to organize the keys under an explicit space name:

```javascript
efemem.set("language:en", "English", "languages");
efemem.set("language:es", "Spanish", "languages");
```



Now, the keys are more easily to remember for search them, because we know the space name in which they are stored.

Remember: you can use complex data as value:

```javascript
efemem.set("student:001", {name: "James", surname: "Gordon", job: "Police inspector"}, "students");
efemem.set("califications", [4.5, 5.5, 6.2, 9, 7.3], "students");
```



If the space and key existis previously, it will be updated with the new value.



### Retrieve a key and value

You can retrieve a value through its key, using the following command:

```javascript
efemem.get("language:en", "languages");
```





### Delete a key and value

You can remove a value through its key, using the following command:

```javascript
efemem.delete("language:en", "languages");
```





### List of spaces

When you are using extensively the database, you will have many keys. Using space names will facilitate you the organization and the search of the keys.

If you want to remember the space names that you are using, simply execute the following command:

```javas
efemem.spaces();
```





### List of keys

You can list all the keys using any of the following command:

```javascript
efemem.keys();
efemem.keys("");
efemem.keys("*");
```



If you have many keys, it will be more easily and efficient specifying the space name:

```javascript
efemem.keys("", "languages");
```



The `keys()` command is very versatile, because the name of the key or the space is used as pattern. This pattern is, simply, a set of characters that are used as part of the complete name. You can use one or more characters.

Imagine you are using the following keys:

`student:001|city:002|course:012`

`student:002|city:002|course:012`

`student:001|city:001|course:007`

`student:010|city:003|course:012`

`student:011|city:003|course:007`



You can use patterns like these:

- `course:012` - Filter keys by course id 012
- `city:003` - Filter keys by city id 003



The same concept can be applied to the space names, or for both. 

For example: imagine you have space names that identify courses for each country (among many others). You will have the following space names: `"coursesUK"`, `"coursesGB"`, `"coursesIT"`, `"coursesFR"` and `"coursesES"`.

The following command will retrieve all the keys for the course ID 002 in all countries:

```javascript
efemem.keys("course:002", "courses");
```



### List of values

You can retrieve the values filtering the keys in the same way you can do with the `keys()` command. 

The `get()` command returns the list of keys, values and space names.



**Examples:**

```javascript
result = efemem.get("", "config");
```



```json
result: {
  "ok": true,
  "cmd": "get(key[,space])",
  "data": [
    {
      "key": "config~maxValue",
      "value": 100
    },
    {
      "key": "config~minValue",
      "value": 1
    }
  ],
  "msg": "2 values found",
  "affected": 2,
  "time": "1ms"
}
```





## More information

**You can read the full documentation at:**

[https://github.com/efememdb/EFEMemDB/blob/master/docs/index.md](https://github.com/efememdb/EFEMemDB/blob/master/docs/index.md)



**History changes at:**

[https://github.com/efememdb/EFEMemDB/blob/master/docs/HISTORY.md](https://github.com/efememdb/EFEMemDB/blob/master/docs/HISTORY.md)

docs/HISTORY.md
# Change log

## 1.0.8 (2020-05-03)

### Features

- Key names by default. If you `set()` a value without a key, a hash value will be assumed as the name of the key. This hash value will contain 24 characters in hexadecimal format, and will assure the key name will be unique and irrepetible.
- Error results are now minimalists: `{ok: false, cmd: "command()", msg: "error message"}`
- Results returns the complete command syntax
- Initial memory size reduced to 10MB
- Improved memory size. `"created"` and `"updated"` fields are removed, saving up to the 50% of memory size.
- Improved pattern keys search performance.
- `delete()` command can remove keys using patterns.
- Removed the following commands:
  - `check()`
  - `values()`
- Improved `get()` command :
  - Use of key patterns
  - `fullInfo` parameter:
    - If `true`, returns the following information:
      - space
      - key
      - data:
        - value
        - due
    - If `true` (by default), returns the following information:
      - space~key
      - value
- Improved `remove()` command:
  - Use of key patterns
  - `overwrite` param:
    - If `true`, when you move a given key to a new space, and this key and space already exists, overwrite this
    - If `false` (by default), when you move a given key to a new space, and this key and space already exists, don't move the key, keeping both keys.
- Improved `copy()` command. Use of key patterns, overwriting existings keys.



### Fixes

- Code more stable, thanks to more exceptions control



## 1.0.7 (2020-04-19)

### Fixes

- When you imported **npm** package in NodeJS, the **efemem** object is not visible. `module.exports` added



## 1.0.6 (2020-04-19)

### Features

- Added `copy()` command
- Added automated testing (`testing` folder)
- Main and unique file: `efememdb.js`
- Code adapted for Web applications
- `example.html` demo
- `stats()` command calculates now statistics with numeric values applying **count**, **sum**, **min**(imum), **max**(imum), **avg** (average), **var** (variance) and **std** (standard deviation) operations



### Fixes

- Added `created`, `updated` and `due` dates to the result of `values()` command
-  `persist()` command refactorized and optimized
- Extension for index file (persistence): `efememdb.efs` (EFemem Spaces)
- Extension for data files (persistence): `<space_name>.efd` (EFemem Data)
- `restore()` command refactorized and optimized
- `move()` repeated an existing key (if this key was recovered from persistence when starting)
- Internal functions `getKeyDetail()` and `deleteIndex()` has been written out of the class, and, henceforth, are not accesible.



## 1.0.4 (2020-04-12)

### Features

- EFEMem DB Logo included in documentation
- Complete code in an unique JS file
- Execution environment identification through the `nodejsEnv` property (`true` or `false`)
- Execution time calculated with `Date()` instead of `process.hrtime()`. This method is valid for NodeJS and standard JavaScript
- Added `info()` command. 



### Fixes

- Documentation corrections
- `keysmax` config parameter changed to `maxKeys`



## 1.0.3 (2020-04-11)

### Fixes

- Main code (efememdb.js) translation to index.js in order to import library correctly
- Set `dataPath` to `'./'` directory by default



## 1.0.2 (2020-04-11)

Tunning package.json



## 1.0.1 (2020-04-11)

Tuninng package.json



## 1.0.0 (2020-04-10)

**EFEMem Database first version.**

### Features

- NodeJS library
- Commands:
  - set
  - get
  - delete
  - rename
  - move
  - check
  - spaces
  - keys
  - values
  - stats
  - memory
  - persist
  - restore
  - getConfig
  - setConfig
- Configuration params:
  - maxMemory
  - maxKeys
  - reclyclingMode
  - dataPath


testing/testing.js
require("./test-set");
require("./test-get");
require("./test-spaces");
require("./test-keys");
require("./test-rename");
require("./test-move");
require("./test-copy");
require("./test-delete");
require("./test-getConfig");
require("./test-setConfig");
require("./test-memory");
require("./test-info");
require("./test-stats");
require("./test-persist");
require("./test-restore");
require("./test-spaceInfo");


testing/test-get.js
const { efemem } = require("../efememdb");

console.log("\n\n*********************");
console.log("Testing command get()");
console.log("*********************");

let ok = 0;
let ko = 0;
let result = {};

/**
 * Evaluates a given result obtained from an executed command
 * @param {string} title => Text that describes the test
 * @param {JSON} result => Result to evaluate
 * @param {boolean} expected => Expected result (true | false)
 */
const evaluate = (title, result, expected) => {
  console.log(`> testing: ${title} - Expected: ${expected}`);

  if (result.ok == expected) {
    ok++;
    console.log(`   [OK] ${result.msg}`);
  } else {
    ko++;
    console.log(`   *** [ERROR] *** ${result.msg}`);
  }
};

// Testing get() without parameters. All keys at 'public' space (2 keys)
result = efemem.get();
evaluate(`get()`, result, true);
console.log(`result: ${JSON.stringify(result, null, 2)}`);

// Testing get() with key null. All keys at 'public' space (2 keys)
result = efemem.get(null);
evaluate(`get(null)`, result, true);

// Testing get(). key 123 at 'public' space not found
result = efemem.get(123);
evaluate(`get(123)`, result, true);

// Testing get() key '1abc' at 'public' space not found
result = efemem.get("1abc");
evaluate(`get("1abc")`, result, true);

// Testing get() key 'maxValue!' at 'public' space not found
result = efemem.get("maxValue!");
evaluate(`get("maxValue!")`, result, true);

// Testing get() a key not found
result = efemem.get(
  "maxValueqweiutyqw573456y23478tyqweiutyqewi57832465897346tr87wqyetiuqweyt2347856234875yt6iequwyteiqwuty34896723"
);
evaluate(
  `get("maxValueqweiutyqw573456y23478tyqweiutyqewi57832465897346tr87wqyetiuqweyt2347856234875yt6iequwyteiqwuty34896723")`,
  result,
  true
);

// Testing get() a key not found
result = efemem.get("maxValue", "1abc");
evaluate(`get("maxValue", "1abc")`, result, true);

// Testing get() a key not found
result = efemem.get("maxValue", "config!");
evaluate(`get("maxValue", "config!")`, result, true);

// Testing get() a key not found
result = efemem.get("maxValue", 100, "configiu09olki87ujhy65tgf");
evaluate(`get("maxValue", 100, "configiu09olki87ujhy65tgf")`, result, true);

// Testing if error. get with an existing key, but not in the 'public' space
result = efemem.get("maxValue");
evaluate(`get("maxValue")`, result, true);

// Testing a non existing key
result = efemem.get("employee:007", "employees");
evaluate(`get("employee:007", "employees")`, result, true);

// Testing an existing key
result = efemem.get("employee:001", "employees");
evaluate(`get("employee:001", "employees")`, result, true);

// Testing an existing key with full info
result = efemem.get("employee:001", "employees", true);
evaluate(`get("employee:001", "employees", true)`, result, true);
console.log(`result: ${JSON.stringify(result, null, 2)}`);

console.log("\n-------------------------------------");
console.log(`OK: ${ok} vs KO: ${ko}`);

if (ko > 0) console.log(`\n\n****** THERE ARE ${ko} ERRORS ******`);
console.log("-------------------------------------");


testing/test-set.js
const { efemem } = require("../efememdb");

console.log("\n\n*********************");
console.log("Testing command set()");
console.log("*********************");

let ok = 0;
let ko = 0;
let result = {};

/**
 * Evaluates a given result obtained from an executed command
 * @param {string} title => Text that describes the test
 * @param {JSON} result => Result to evaluate
 * @param {boolean} expected => Expected result (true | false)
 */
const evaluate = (title, result, expected) => {
  console.log(`> testing: ${title} - Expected: ${expected}`);

  if (result.ok == expected) {
    ok++;
    console.log(`   [OK] ${result.msg}`);
  } else {
    ko++;
    console.log(`   *** [ERROR] *** ${result.msg}`);
  }
};

// Testing simple "hello" "world" in 'public' space
result = efemem.set("hello", "world");
evaluate(`set("hello", "world)`, result, true);

// Testing if error. set() without parameters (no value provided)
result = efemem.set();
evaluate(` efemem.set()`, result, false);

// Testing if error. set() without value
result = efemem.set("maxValue");
evaluate(`set("maxValue")`, result, false);

// Testing if error. set() with incorrect key (null). By default, key hash
result = efemem.set(null, 100);
evaluate(`set(null, 100)`, result, true);
console.log(`result: ${JSON.stringify(result, null, 2)}`);
// Testing if error. set() with incorrect key (no string name)
result = efemem.set(123, 100);
evaluate(`set(123, 100)`, result, false);

// Testing if error. set() with incorrect key (first character not alphabetical)
result = efemem.set("1abc", 100);
evaluate(`set("1abc", 100)`, result, false);

// Testing if error. set() with incorrect key (illegal character)
result = efemem.set("maxValue!", 100);
evaluate(`set("maxValue!", 100)`, result, false);

// Testing if error. set() with incorrect key (More than 100 characters)
result = efemem.set(
  "maxValueqweiutyqw573456y23478tyqweiutyqewi57832465897346tr87wqyetiuqweyt2347856234875yt6iequwyteiqwuty34896723",
  100
);
evaluate(
  `set("maxValueqweiutyqw573456y23478tyqweiutyqewi57832465897346tr87wqyetiuqweyt2347856234875yt6iequwyteiqwuty34896723", 100)`,
  result,
  false
);

// Testing if error. set with incorrect space (first character not alphabetical)
result = efemem.set("maxValue", 100, "1abc");
evaluate(`set("maxValue", 100, "1abc")`, result, false);

// Testing if error. set with incorrect space (illegal character)
result = efemem.set("maxValue", 100, "config!");
evaluate(`set("maxValue", 100, "config!")`, result, false);

// Testing if error. set with incorrect space (more than 24 characters)
result = efemem.set("maxValue", 100, "configiu09olki87ujhy65tgf");
evaluate(`set("maxValue", 100, "configiu09olki87ujhy65tgf")`, result, false);

// Testing set(). key="maxValue", value=100. space="config"  (integer value)
result = efemem.set("maxValue", 100, "config");
evaluate(`set("maxValue", 100, "config")`, result, true);

// Testing set(). key="salary", value=1234.56. space="employees"  (float value)
result = efemem.set("salary", 1234.56, "employees");
evaluate(`set("salary", 1234.56, "employees")`, result, true);

// Testing set(). key="job", value="IT Developer". space="employees"  (text value)
result = efemem.set("job", "IT Developer", "employees");
evaluate(`set("job", "IT Developer", "employees")`, result, true);

// Testing set(). key="cacheUsage", value=false, space="config"  (boolean value)
result = efemem.set("cacheUsage", false, "config");
evaluate(`set("cacheUsage", false, "config")`, result, true);

// Testing set() with an array value
result = efemem.set(
  "languages",
  ["NodeJS", "JavaScript", "Go", "Python", "C++"],
  "employees"
);
evaluate(`set() with array`, result, true);

// Testing set() with a JSON object
result = efemem.set(
  "employee:001",
  { name: "Rafael", surname: "Hernamperez", position: "CTO" },
  "employees"
);
evaluate(`set() with JSON object`, result, true);

console.log("\n-------------------------------------");
console.log(`OK: ${ok} vs KO: ${ko}`);

if (ko > 0) console.log(`\n\n****** THERE ARE ${ko} ERRORS ******`);
console.log("-------------------------------------");


index.js
/*
/\/\/\/\/\/\/\/\/\/\/\/\/\/\/\/\/\/\/\/\/\/\/\/\/\/\/\/\/\/\/\/\/\/\/\/\/\/\/\
index.js

Easy, Fast and Efficient MEMory NoSQL DataBase
Version 1.0.7

Created by Rafael Hernamperez and released under the terms of the ISC License:
https://opensource.org/licenses/ISC



ISC License (ISC)
Copyright 2020 Rafael Hernamperez

Permission to use, copy, modify, and/or distribute this software for any 
purpose with or without fee is hereby granted, provided that the above 
copyright notice and this permission notice appear in all copies.

THE SOFTWARE IS PROVIDED "AS IS" AND THE AUTHOR DISCLAIMS ALL WARRANTIES WITH
REGARD TO THIS SOFTWARE INCLUDING ALL IMPLIED WARRANTIES OF MERCHANTABILITY 
AND FITNESS. IN NO EVENT SHALL THE AUTHOR BE LIABLE FOR ANY SPECIAL, DIRECT, 
INDIRECT, OR CONSEQUENTIAL DAMAGES OR ANY DAMAGES WHATSOEVER RESULTING FROM 
LOSS OF USE, DATA OR PROFITS, WHETHER IN AN ACTION OF CONTRACT, NEGLIGENCE OR 
OTHER TORTIOUS ACTION, ARISING OUT OF OR IN CONNECTION WITH THE USE OR 
PERFORMANCE OF THIS SOFTWARE.
\/\/\/\/\/\/\/\/\/\/\/\/\/\/\/\/\/\/\/\/\/\/\/\/\/\/\/\/\/\/\/\/\/\/\/\/\/\/\/
*/
const { efemem } = require("./efememdb");

module.exports = { efemem };


examples.js
/*
/\/\/\/\/\/\/\/\/\/\/\/\/\/\/\/\/\/\/\/\/\/\/\/\/\/\/\/\/\/\/\/\/\/\/\/\/\/\/\
examples.js

EFEMem NoSQL DataBase examples
Version 1.0.4

Created by Rafael Hernamperez and released under the terms of the ISC License:
https://opensource.org/licenses/ISC

ISC License (ISC)
Copyright 2020 Rafael Hernamperez

\/\/\/\/\/\/\/\/\/\/\/\/\/\/\/\/\/\/\/\/\/\/\/\/\/\/\/\/\/\/\/\/\/\/\/\/\/\/\/
*/
const { efemem } = require("efememdb");

// EFEMem Database version
console.log(efemem.info());

// Environment
if (!efemem.webEnv) console.log(`EFEMem DB is running on NodeJS`);
else console.log(`EFEMem DB is running out of NodeJS`);

// Data assignation
// Integer value
let result = efemem.set("maxValue", 100, "config");
// Float value
result = efemem.set("salary", "1234.56", "employees");
// Text value
result = efemem.set("job", "IT Developer", "employees");
// Boolean value
result = efemem.set("cacheUsage", false, "config");
// Array value
result = efemem.set(
  "languages",
  ["NodeJS", "JavaScript", "Go", "Python", "C++"],
  "employees"
);
// JSON object value
result = efemem.set(
  "employee:001",
  { name: "Rafael", surname: "Hernamperez", position: "CTO" },
  "employees"
);

// Get data
result = efemem.get("employee:001", "employees");
console.log(`Employee: ${JSON.stringify(result, null, 2)}`);

// Spaces list
result = efemem.spaces();
console.log(`Spaces: ${JSON.stringify(result, null, 2)}`);

// Keys list
result = efemem.keys(); // All the keys
console.log(`Keys: ${JSON.stringify(result, null, 2)}`);
result = efemem.keys("", "config"); // All the keys from 'config' space
console.log(`'config' keys: ${JSON.stringify(result, null, 2)}`);

// Get values of the keys from '*emp*' space (employees)
result = efemem.values("", "emp");
console.log(`'*emp*' values: ${JSON.stringify(result, null, 2)}`);

// rename 'employee:001' to 'employee:007'
result = efemem.rename("employee:001", "employee:007", "employees");
console.log(`'Rename:' values: ${JSON.stringify(result, null, 2)}`);

// Move key 'salary' to 'public' space
result = efemem.move("salary", "employees");
console.log(`Move: ${JSON.stringify(result, null, 2)}`);
result = efemem.values("", "public");
console.log(`'public' values: ${JSON.stringify(result, null, 2)}`);

// Delete key 'job' from 'employees' space
result = efemem.delete("job", "employees");
console.log(`Delete: ${JSON.stringify(result, null, 2)}`);

// Final Data
console.log("----------");
console.log("FINAL DATA");
console.log("----------");
result = efemem.spaces();
console.log(`Spaces: ${JSON.stringify(result, null, 2)}`);
result = efemem.keys(); // All the keys
console.log(`Keys: ${JSON.stringify(result, null, 2)}`);
result = efemem.values();
console.log(`values: ${JSON.stringify(result, null, 2)}`);
console.log(`info: ${JSON.stringify(efemem.info(), null, 2)}`);

// Delete key 'job' from 'employees' space
result = efemem.memory();
console.log(`Memory: ${JSON.stringify(result, null, 2)}`);

// Statistics
result = efemem.stats();
console.log(`Statistics: ${JSON.stringify(result, null, 2)}`);


examples/nodejs-example.js
/*
/\/\/\/\/\/\/\/\/\/\/\/\/\/\/\/\/\/\/\/\/\/\/\/\/\/\/\/\/\/\/\/\/\/\/\/\/\/\/\
examples.js

EFEMem NoSQL DataBase example for NodeJS
Version 1.0.7

Created by Rafael Hernamperez and released under the terms of the ISC License:
https://opensource.org/licenses/ISC

ISC License (ISC)
Copyright 2020 Rafael Hernamperez

\/\/\/\/\/\/\/\/\/\/\/\/\/\/\/\/\/\/\/\/\/\/\/\/\/\/\/\/\/\/\/\/\/\/\/\/\/\/\/
*/
const { efemem } = require("../efememdb");

// EFEMem Database version
console.log(efemem.info());

// Environment
if (!efemem.webEnv) console.log(`EFEMem DB is running on NodeJS`);
else console.log(`EFEMem DB is running out of NodeJS`);

// Data assignation
// Integer value
let result = efemem.set("maxValue", 100, "config");
// Float value
result = efemem.set("salary", 1234.56, "employees");
// Text value
result = efemem.set("job", "IT Developer", "employees");
// Boolean value
result = efemem.set("cacheUsage", false, "config");
// Array value
result = efemem.set(
  "languages",
  ["NodeJS", "JavaScript", "Go", "Python", "C++"],
  "employees"
);
// JSON object value
result = efemem.set(
  "employee:001",
  { name: "Rafael", surname: "Hernamperez", position: "CTO" },
  "employees"
);

// Get data
result = efemem.get("employee:001", "employees");
console.log(`Employee: ${JSON.stringify(result, null, 2)}`);

// Spaces list
result = efemem.spaces();
console.log(`Spaces: ${JSON.stringify(result, null, 2)}`);

// Keys list
result = efemem.keys(); // All the keys
console.log(`Keys: ${JSON.stringify(result, null, 2)}`);
result = efemem.keys("", "config"); // All the keys from 'config' space
console.log(`'config' keys: ${JSON.stringify(result, null, 2)}`);

// Get values of the keys from '*emp*' space (employees)
result = efemem.get("", "emp");
console.log(`'*emp*' values: ${JSON.stringify(result, null, 2)}`);

// rename 'employee:001' to 'employee:007'
result = efemem.rename("employee:001", "employee:007", "employees");
console.log(`'Rename:' values: ${JSON.stringify(result, null, 2)}`);

// Move key 'salary' to 'public' space
result = efemem.move("salary", "employees");
console.log(`Move: ${JSON.stringify(result, null, 2)}`);
result = efemem.get("", "public");
console.log(`values: ${JSON.stringify(result, null, 2)}`);

// Copy key 'salary' from 'public' space to 'employees' space
result = efemem.copy("salary", "public", "salary", "employees");
console.log(`copy: ${JSON.stringify(result, null, 2)}`);
result = efemem.get("salary");
console.log(`'keys "salary"' values: ${JSON.stringify(result, null, 2)}`);

// Delete key 'job' from 'employees' space
result = efemem.delete("job", "employees");
console.log(`Delete: ${JSON.stringify(result, null, 2)}`);

// Final Data
console.log("----------");
console.log("FINAL DATA");
console.log("----------");
result = efemem.spaces();
console.log(`Spaces: ${JSON.stringify(result, null, 2)}`);
result = efemem.keys(); // All the keys
console.log(`Keys: ${JSON.stringify(result, null, 2)}`);
result = efemem.get();
console.log(`values: ${JSON.stringify(result, null, 2)}`);
console.log(`info: ${JSON.stringify(efemem.info(), null, 2)}`);

// Delete key 'job' from 'employees' space
result = efemem.memory();
console.log(`Memory: ${JSON.stringify(result, null, 2)}`);

// Persist data
result = efemem.persist();
console.log(`persist: ${JSON.stringify(result, null, 2)}`);

// Restore data
result = efemem.restore();
console.log(`restore: ${JSON.stringify(result, null, 2)}`);

// Values
result = efemem.get();
console.log(`values: ${JSON.stringify(result, null, 2)}`);

// Prepare data for statistics
efemem.set("student:001|class:Python", 9.7, "qualifications");
efemem.set("student:002|class:Python", 7.5, "qualifications");
efemem.set("student:003|class:Python", 6.7, "qualifications");
efemem.set("student:001|class:JavaScript", 8.7, "qualifications");
efemem.set("student:002|class:JavaScript", 6.6, "qualifications");
efemem.set("student:003|class:JavaScript", 8.3, "qualifications");
efemem.set("student:001|class:Java", "9.2", "qualifications");
efemem.set("student:002|class:Java", "9.5", "qualifications");
efemem.set("student:003|class:Java", "8.9", "qualifications");
efemem.set("student:005|class:Python", 5.8, "qualifications");

// Statistics for Python
result = efemem.stats("Python");
console.log(`Python Statistics: ${JSON.stringify(result, null, 2)}`);

// Statistics for class Java and JavaScript
result = efemem.stats("Java", "qualifications");
console.log(`Java Statistics: ${JSON.stringify(result, null, 2)}`);

// Statistics for student 001
result = efemem.stats("student:001", "qualifications");
console.log(`Student 001 Statistics: ${JSON.stringify(result, null, 2)}`);

// Space information
result = efemem.spaceInfo("qualifications");
console.log(`space 'qualifications' info: ${JSON.stringify(result, null, 2)}`);

// Persist data
result = efemem.persist();
console.log(`persist: ${JSON.stringify(result, null, 2)}`);
